CMakeLists.txt

```diff
@@ -0,0 +1,59 @@
+cmake_minimum_required(VERSION 2.8)
+
+include(GNUInstallDirs)
+
+find_package(LLVM REQUIRED CONFIG)
+message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
+message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
+
+add_definitions(${LLVM_DEFINITIONS})
+include_directories(${LLVM_INCLUDE_DIRS})
+link_directories(${LLVM_LIBRARY_DIRS})
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti")
+#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+
+add_definitions(-D_GNU_SOURCE
+	-DKLEE_DIR=\"${CMAKE_BINARY_DIR}\"
+	-DKLEE_INSTALL_BIN_DIR=\"${CMAKE_INSTALL_FULL_BINDIR}\"
+	-DKLEE_INSTALL_RUNTIME_DIR=\"${CMAKE_INSTALL_FULL_LIBDIR}/klee/runtime\")
+
+set(PACKAGE_STRING "\"KLEE 1.2.0\"")
+set(PACKAGE_URL "\"https://klee.github.io\"")
+set(HAVE_GPERFTOOLS_MALLOC_EXTENSION_H 1)
+set(ENABLE_STP 1)
+set(HAVE_MALLINFO 1)
+set(HAVE_MEMORY_H 1)
+set(KLEE_CMAKE_BUILD 1)
+set(RUNTIME_CONFIGURATION "\"\"")
+
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/include/klee/Config/config.h.cmin
+	${CMAKE_CURRENT_BINARY_DIR}/include/klee/Config/config.h)
+
+include_directories(include)
+
+macro(klee_build_llvm_library lib)
+	foreach(in_f ${ARGN})
+		get_filename_component(out_f ${in_f} NAME)
+		set(out_f ${out_f}.bc)
+		add_custom_command(OUTPUT ${out_f}
+			COMMAND ${CLANG_PATH} -emit-llvm -c -o ${out_f} ${in_f}
+			DEPENDS ${in_f}
+			COMMENT "Building LLVM bitcode ${out_f}"
+			VERBATIM
+		)
+		list(APPEND result ${out_f})
+	endforeach()
+	add_custom_command(OUTPUT ${lib}.bc
+		COMMAND ${LLVM_LINK_PATH} -o ${lib}.bc ${result}
+		DEPENDS ${result}
+		COMMENT "Linking LLVM bitcode ${lib}.bc"
+		VERBATIM
+	)
+	add_custom_target(${lib} ALL DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${lib}.bc)
+endmacro()
```

Comment:
This is the part of KLEE I knew would always be a problem. The problem with your above approach is that it does not handle dependencies properly. For example if a header file that these sources depends on changes invoking the build system will not rebuild the bitcode file even though it should.

I've thought about this. Even though `add_custom_command(...)` has the `IMPLICIT_DEPENDS` option which can use to fix this, it only works for Makefile builds and is totally broken for other generators. The right way to fix this I believe is to treat KLEE's runtimes as an external project and build using `ExternalProject()`. We would have a separate build system for the runtime. We could use CMake for this as well but I think I'd like to avoid this because getting CMake to build the bitcode we want is kind of difficult without using a wrapper like whole-program-llvm to do the work.